Wasm/Engine.lean

```diff
@@ -180,139 +165,152 @@ def findLocalByName? (ls : List (Option String × Option StackEntry))
       | .some name => match findLocalByName? locals name with
-        | .none => .error $ .local_with_given_name_missing name
-        | .some l => .ok ((stack, .none), l)
+        | .none => throwEE $ .local_with_given_name_missing name
+        | .some l => pure l
     | .by_index i => match locals.get? i.index with
-      | .some (_, .some se)  => .ok ((stack, .none), se)
-      | _ => .error $ .local_with_given_id_missing i.index
+      | .some (_, .some se) => pure se
+      | _ => throwEE $ .local_with_given_id_missing i.index
 
-  partial def runWithContinuation
+  partial def computeContinuation
                       (blocktypes : List Type')
                       (locals : List (Option String × Option StackEntry))
-                      (so : ContinuationStack)
                       (ops' : List Operation)
-                      : Except EngineErrors (List StackEntry) := do
+                      : EngineM PUnit := do
     let rec go
-    | (stack, .none), [] => pure stack
-    | (stack, .some cont), _ => go (stack, .none) cont
-    | (stack, .none), op :: ops => do go (←runOp locals stack op) ops
+    | [] => pure ()
+    | op :: ops => do match ←(runOp locals op).run (←get) with
+      | (.error cont, stack') => set stack'; go cont
+      | (.ok _, stack') => set stack'; go ops
 
-    let es' ← (stackValues ∘ Stack.mk) <$> go so ops'
+    go ops'
+    let es' := stackValues ⟨←get⟩
     if resultsTypecheck blocktypes es'
-      then pure es'
-      else throw .stack_incompatible_with_results
+      then set es'
+      else throwEE .stack_incompatible_with_results
 
-  -- TODO: this `getSO` threading shit literally simulates a monad in a way.
-  -- We NEED to rewrite continuations to be passed more sanely.
-  -- (`ContinuationStack` monad? `Except` pulling double-weight?)
-  partial def runIBinop (locals : List (Option String × Option StackEntry))
-                        (stack : List StackEntry)
-                        (g0 : Get') (g1 : Get')
-                        (binop : Int → Int → Int)
-                        : Except EngineErrors ContinuationStack := do
-    let ((stack', cont?), operand0) ← getSO locals stack g0
-    if cont?.isSome then pure (operand0 :: stack', cont?) else
-      let ((stack1, cont1?), operand1) ← getSO locals stack' g1
-      if cont1?.isSome then pure (operand1 :: stack1, cont1?) else
-        let res ← match operand0, operand1 with
-          -- TODO: check bitsize and overflow!
-          | .num (.i ⟨b0, i0⟩), .num (.i ⟨_b1, i1⟩) =>
-              pure $ .num $ .i ⟨b0, binop i0 i1⟩
-          | _, _ => throw .param_type_incompatible
-        pure (res :: stack1, .none)
-
-  -- TODO: there's a StateT somewhere here. Just sayin'
-  -- TODO: we're not typechecking at all!
-  -- TODO: can locals change when executing nested instructions?
+  -- TODO: check that typechecking is done everywhere!
   partial def runOp (locals : List (Option String × Option StackEntry))
```

Comment:
I didn't review the logic, but I've reviewed the overall idea and it seems correct.